src/core/include/units/math.h

```diff
@@ -147,4 +147,43 @@ template<Quantity Q>
   return Q(std::numeric_limits<typename Q::rep>::epsilon());
 }
 
+/**
+ * @brief Returns the greatest quantity that is representable in Unit To that is less or equal to q.
+ *
+ * The returned value is rounded down to a quantity with unit type To
+ *
+ * @tparam q Quantity type being the base of the operation
+ * @return Quantity The rounded quantity with unit type To
+ */
+template<Unit To, typename D, typename U, typename Rep>
+[[nodiscard]] constexpr quantity<D, To, Rep> floor(const quantity<D, U, Rep>& q) noexcept
+  requires (!treat_as_floating_point<Rep>) ||
+    requires { floor(q.number()); } ||
+    requires { std::floor(q.number()); }
+{
+  auto handle_signed_results = [&q](auto&& res) constexpr
```

Comment:
I do not see the benefit of passing the argument here as a forwarding reference (`std::forward` is not used) and it only complicates stuff as you have to add `std::remove_reference_t<T>`.

`const` lvalue reference should be fine.

Reply:
I agree for the `const` lvalue reference, but I guess we still need `std::remove_reference_t<T>`. Because as specified `res` is a `const` lvalue reference which does not have the `static one()`.

Reply:
Right, sorry I was more thinking about something like that:

```cpp
auto handle_signed_results = [&]<typename T>(const T& res) {
  if (res > q)
    return res - T::one();
  return res;
};
```

Also, please do not type `[&q]`. `[&]` is enough and we should not help the compiler.

Reply:
I used `[&q]` according to Scott Meyers' _Modern Effective C++, Item 31: Avoid default capture modes_ to avoid dangling references. This is of course just a runtime issue, but I tried to be consitent. What do you think?

Reply:
In general, we always say "do not help the compiler" :-) Most of those end up as maintenance issues sooner or later.

Reply:
The closure doesn't outlive the function, so there's no danger of dangling references. The closure should be const, though.